heron/packing/src/java/com/twitter/heron/packing/PackingPlanBuilder.java

```diff
@@ -0,0 +1,170 @@
+// Copyright 2016 Twitter. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.twitter.heron.packing;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Logger;
+
+import com.google.common.base.Optional;
+
+import com.twitter.heron.spi.packing.InstanceId;
+import com.twitter.heron.spi.packing.PackingPlan;
+import com.twitter.heron.spi.packing.Resource;
+
+/**
+ * Class the help with building packing plans.
+ */
+public class PackingPlanBuilder {
+  private static final Logger LOG = Logger.getLogger(PackingPlanBuilder.class.getName());
+
+  private final String topologyId;
+  private final PackingPlan existingPacking;
+  private Resource defaultInstanceResource;
+  private Resource maxContainerResource;
+  private Map<String, Long> componentRamMap;
+  private int requestedContainerPadding;
+  private int numContainers;
+
+  private Map<Integer, List<InstanceId>> containerInstances;
+  private List<Container> containers;
+
+  public PackingPlanBuilder(String topologyId) {
+    this(topologyId, null);
+  }
+
+  public PackingPlanBuilder(String topologyId, PackingPlan existingPacking) {
+    this.topologyId = topologyId;
+    this.existingPacking = existingPacking;
+    this.containerInstances = new HashMap<Integer, List<InstanceId>>();
+    this.numContainers = 0;
+  }
+
+  // set resource settings
+  public PackingPlanBuilder setDefaultInstanceResource(Resource resource) {
+    this.defaultInstanceResource = resource;
+    return this;
+  }
+
+  public PackingPlanBuilder setMaxContainerResource(Resource resource) {
+    this.maxContainerResource = resource;
+    return this;
+  }
+
+  public PackingPlanBuilder setRequestedComponentRam(Map<String, Long> ramMap) {
+    this.componentRamMap = ramMap;
+    return this;
+  }
+
+  public PackingPlanBuilder setRequestedContainerPadding(int percent) {
+    this.requestedContainerPadding = percent;
+    return this;
+  }
+
+  public PackingPlanBuilder updateNumContainers(int count) {
```

Comment:
I'm a little confused here. The command to add a container is the following: PackingUtils.allocateNewContainer(containers, containers.get(0).getCapacity(), this.requestedContainerPadding) when someone adds a new container. Why can't we do this in the updatecommand? It doesn't have any connection to an instance.

Overall, what makes it hard for me to verify the correctness of the code is that the numContainers in the builder can change without allocating the necessary containers right away. Wouldn't it be easier to enforce that by requiring that when the numContainers is updated in the builder, the List<Containers> is also updated in a "transaction".